core/lib/zksync_core/src/api_server/web3/backend_jsonrpsee/middleware.rs

```diff
@@ -145,27 +147,42 @@ where
             .copied()
             .unwrap_or("");
 
-        WithMethodCall {
-            call: self.method_tracer.new_call(method_name),
-            inner: self.inner.call(request),
-        }
+        let observed_params = ObservedRpcParams::new(request.params.as_ref());
+        let call = self.method_tracer.new_call(method_name, observed_params);
+        WithMethodCall::new(self.inner.call(request), call)
     }
 }
 
 pin_project! {
     #[derive(Debug)]
-    pub(crate) struct WithMethodCall<F> {
-        call: MethodCall,
+    pub(crate) struct WithMethodCall<'a, F> {
         #[pin]
         inner: F,
+        call: MethodCall<'a>,
+        call_span: tracing::Span,
     }
 }
 
-impl<F: Future<Output = MethodResponse>> Future for WithMethodCall<F> {
+impl<'a, F> WithMethodCall<'a, F> {
+    fn new(inner: F, call: MethodCall<'a>) -> Self {
+        // Wrap a call into a span with unique correlation ID, so that events occurring in the span can be easily filtered.
+        // This works as a cheap alternative to Open Telemetry tracing with its trace / span IDs.
+        let call_span =
+            tracing::debug_span!("rpc_call", correlation_id = generate_correlation_id());
```

Comment:
I've used this trick on my previous job (although there, it worked with conjunction with Jaeger tracing), and it really helped to investigate some errors. It is another allocation per each request, but it's small.